libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -760,12 +761,18 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 
 	if (functionCallKind == FunctionCallKind::TypeConversion)
 	{
-		solAssert(
-			_functionCall.expression().annotation().type->category() == Type::Category::TypeType,
-			"Expected category to be TypeType"
-		);
 		solAssert(_functionCall.arguments().size() == 1, "Expected one argument for type conversion");
-		define(_functionCall, *_functionCall.arguments().front());
+		Type const* fromType = _functionCall.arguments().front()->annotation().type;
+		Type const* toType = _functionCall.expression().annotation().type;
+
+		solAssert(toType->category() == Type::Category::TypeType, "Expected category to be TypeType");
+		Type const* toActualType = dynamic_cast<TypeType const*>(toType)->actualType();
+
+		if (fromType->category() == Type::Category::TypeType && toActualType->category() == Type::Category::Address)
+			define(_functionCall) << linkerSymbol(dynamic_cast<TypeType const&>(*fromType)) << "\n";
```

Comment:
Can you add some more assertions here? I know they are inline `linkerSymbol` but this looks a bit weird - why would any TypeType conversion to address lead to linker Symbol?

Reply:
I'm actually just looking at `YulUtilFunctions::conversionFunction`. The checks can be handled there - why would you need to do it here?

Reply:
Doing it in `YulUtilFunctions::conversionFunction()` was my first implementation but then I ran into errors about invalid IR. Same thing you can see in the failing tests after your changes: https://app.circleci.com/pipelines/github/ethereum/solidity/10138/workflows/a6059bc2-7bae-408f-8dfb-6f92cc699373/jobs/499676

```yul
            function fun_g_61(vloc_a_36, vloc_b_38) -> vloc__41 {
                let zero_value_for_type_t_bool_2 := zero_value_for_split_t_bool()
                vloc__41 := zero_value_for_type_t_bool_2

                let expr_48 := convert_t_type$_t_contract$_L_$18_$_to_t_address(expr_47_address)
                let expr_49_address := expr_48
                ...
```
```
Invalid IR generated:
:178:81: Error: Identifier not found.
                let expr_26 := convert_t_type$_t_contract$_L_$18_$_to_t_address(expr_25_address)
                                                                                ^-------------^
:191:81: Error: Identifier not found.
                let expr_48 := convert_t_type$_t_contract$_L_$18_$_to_t_address(expr_47_address)
                                                                                ^-------------^
```

The thing is that this is not really a conversion. A conversion function requires a parameter but there's nothing to give it. Well, I could give it the address but then why call it at all? I could also create a dummy variable set to 0 and ignore the parameter completely inside the function but the current solution of just skipping the conversion seemed cleaner to me.

Reply:
> Can you add some more assertions here? I know they are inline `linkerSymbol` but this looks a bit weird - why would any TypeType conversion to address lead to linker Symbol?

How about a comment?